### PR TITLE
Skip attestations that cannot be verified rather than abort

### PR DESCRIPTION
**Pull request by adityasaky**
If there are extra attestations signed by keys not in the policy, the verifier currently terminates with an error, even if the attestation isn't necessary for the layout. This PR makes the verifier ignore such claims, proceeding with verification of authenticated attestations.

### CODE REVIEW

File: verifier/verifier.go [hunk 1]
```diff
@@ -59,7 +59,8 @@ func Verify(layout *Layout, attestations map[string]*dsse.Envelope, parameters m
 
 		acceptedKeys, err := envVerifier.Verify(context.Background(), env)
 		if err != nil {
-			return err
+			log.Infof("Unable to verify %s's signatures", attestationName)
```
Comment (trishankatdatadog) [inline verifier/verifier.go:62]: Looks safe... I think. As long as we handle all of the following cases, we will be OK:
1. Attestations for known steps
    1. Attestations for known keys
        1. Crossing the threshold
    3. Attestations for unknown keys
2. Attestations for unknown steps
    1. Attestations for known keys
    2. Attestations for unknown keys

I don't like mixing 1.1 and 2.1 as a preloaded check here. IMHO, we shouldn't even be considering (2) at all, but this requires a deeper refactoring.

Comment (adityasaky) [reply verifier/verifier.go:62]: Deeper refactoring is a different question. In this verifier, I explicitly set out to differentiate claim authentication and claim verification. So all claims are loaded and authenticated into a format of "X says C for step S", and policy verification is for the question "does one of {X, Y, Z} say C for step S?"

Comment (adityasaky) [reply verifier/verifier.go:62]: I'm open to the other approach where we use the policy to find what attestations we load and authenticate. The approach I was using seemed to also line up with having, say, a bundle of attestations that weren't on disk or so.

Comment (marcelamelara) [reply verifier/verifier.go:62]: Interesting discussion! I can see advantages to both approaches (using the policy to find attestations vs separate auth from verification). The former more tightly binds the attestations to the policy, and avoids checking irrelevant attestations, though I don't think that precludes retrieving attestations from a remote location. That info could be added to the layout. 

What I think is nice about the latter approach (the current design) is that it does give us more flexibility in terms of policy expression and verification. It lends itself more to expressing policy as "I want n of m trusted attestors to claim C about subject A during step S" and speculatively check an entire (set of) bundle. I'm trying to see a security risk in trying to _authenticate_ attestations for unknown signers (and steps for that matter), but can't immediately think of anything that jumps out right away.

Comment (adityasaky) [reply verifier/verifier.go:62]: nb: we're not authenticating unknown signers after all, because the layout determines known signers.

**Review by marcelamelara — APPROVED** [review]
Given our discussion, I'm fine with accepting this. My only request prior to merging would be to document the main points of the discussion in a comment here for posterity's sake.

**Comment by adityasaky** [comment]
@marcelamelara https://github.com/in-toto/attestation-verifier/pull/18/commits/69b8d6236f830a4803b94a0d4ea268a973e5e983

**Review by marcelamelara — DISMISSED** [review]
LGTM, thank you @adityasaky !

**Comment by adityasaky** [comment]
@trishankatdatadog feel free to merge if your concern's been addressed, thanks!

File: verifier/verifier.go [hunk 2]
```diff
@@ -59,7 +59,19 @@ func Verify(layout *Layout, attestations map[string]*dsse.Envelope, parameters m
...
+			// one of <trusted signers> say <claim> for <step>?"
+			// While this might result in verifying the signatures of
+			// attestations that aren't required for the specific layout, it
+			// more cleanly separates policy evaluation from claim expression.
+			// Also, we do not authenticate attestations from unknown verifiers,
+			// we only hit a situation here where trusted keys from a trusted
+			// layout cannot authenticate an attestation, which is discarded and
+			// not used in the rest of the verification workflow.
+			log.Infof("Unable to verify %s's signatures", attestationName)
+			continue
```
Comment (trishankatdatadog) [inline verifier/verifier.go:74]: How is it discarded? That's not obvious to me right now.

Comment (adityasaky) [reply verifier/verifier.go:75]: updated text

**Review by marcelamelara — APPROVED** [review]
Thanks for the updates!

**Review by trishankatdatadog — APPROVED** [review]
This seems to work for now, but it took me reading the code twice to understand why it apparently does. I understand this verifier is a prototype for now, but I think it needs a significant refactoring to ensure that mistakes are easier to spot and harder to make. Thanks!

**Pull request merged**